redwood-lazylayout-view/src/main/kotlin/app/cash/redwood/lazylayout/view/ViewLazyList.kt

```diff
@@ -84,22 +94,62 @@ internal class ViewLazyList(
         }
       },
     )
+    value.addOnScrollListener(
+      object : RecyclerView.OnScrollListener() {
+        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
+          updateViewport()
+        }
+
+        override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
+          updateViewport()
+        }
+      },
+    )
   }
 
   override fun isVertical(isVertical: Boolean) {
     linearLayoutManager.orientation = if (isVertical) RecyclerView.VERTICAL else RecyclerView.HORIZONTAL
   }
 
-  override fun onPositionDisplayed(onPositionDisplayed: (Int) -> Unit) {
-    adapter.onPositionDisplayed = onPositionDisplayed
+  override fun onViewportChanged(onViewportChanged: (IntArray) -> Unit) {
+    this.onViewportChanged = onViewportChanged
   }
 
-  private class LazyContentItemListAdapter : RecyclerView.Adapter<ViewHolder>() {
-    var onPositionDisplayed: ((Int) -> Unit)? = null
+  private fun updateViewport() {
+    val newViewport = linearLayoutManager.findFirstVisibleItemPosition()..linearLayoutManager.findLastVisibleItemPosition()
+    if (newViewport != viewport) {
+      this@ViewLazyList.viewport = newViewport
+      onViewportChanged?.invoke(intArrayOf(newViewport.first, newViewport.last + 1))
```

Comment:
Do we do too many `+ 1` calls on the range end? I think if we have the `+ 1` here we can probably remove it on the assignment of `itemsAfter` in `LazyList.kt`.

Reply:
Checked the math, and I’m wrong. As-is is correct! (I really wish they hadn’t deprecated `endInclusive`. What a fail.)

Reply:
I added some logs and `linearLayoutManager.findLastVisibleItemPosition() == newViewport.last`, so I'm gonna remove the `+ 1` and propagate that to the callback.